### ISSUE TITLE
error while exporting XGBoostClassifier to PMML

### ISSUE THREAD

**Issue by emilpop-au** [jpmml/jpmml-xgboost#36]
Hello Villu,

I'm having an issue with exporting an XGBoostClassifier model to PMML.  The objective function is 'binary:logistic' and the dependent variable takes 0/1 values.

When exporting the pipeline that contains only 1 step (the xgb trained model) I get the following error:

--------------------------------------------------------------------------
Standard output is empty
Standard error:
May 01, 2019 1:06:35 PM org.jpmml.sklearn.Main run
INFO: Parsing PKL..
May 01, 2019 1:06:35 PM org.jpmml.sklearn.Main run
SEVERE: Failed to parse PKL
java.lang.IllegalArgumentException: |u1
	at numpy.core.NDArrayUtil$TypeDescriptor.read(NDArrayUtil.java:564)
	at numpy.core.NDArrayUtil.parseArray(NDArrayUtil.java:215)
	at numpy.core.NDArrayUtil.parseData(NDArrayUtil.java:190)
	at joblib.NumpyArrayWrapper.toArray(NumpyArrayWrapper.java:43)
	at org.jpmml.sklearn.PickleUtil$1.dispatch(PickleUtil.java:88)
	at net.razorvine.pickle.Unpickler.load(Unpickler.java:122)
	at org.jpmml.sklearn.PickleUtil.unpickle(PickleUtil.java:98)
	at org.jpmml.sklearn.Main.run(Main.java:104)
	at org.jpmml.sklearn.Main.main(Main.java:94)

Exception in thread "main" java.lang.IllegalArgumentException: |u1
	at numpy.core.NDArrayUtil$TypeDescriptor.read(NDArrayUtil.java:564)
	at numpy.core.NDArrayUtil.parseArray(NDArrayUtil.java:215)
	at numpy.core.NDArrayUtil.parseData(NDArrayUtil.java:190)
	at joblib.NumpyArrayWrapper.toArray(NumpyArrayWrapper.java:43)
	at org.jpmml.sklearn.PickleUtil$1.dispatch(PickleUtil.java:88)
	at net.razorvine.pickle.Unpickler.load(Unpickler.java:122)
	at org.jpmml.sklearn.PickleUtil.unpickle(PickleUtil.java:98)
	at org.jpmml.sklearn.Main.run(Main.java:104)
	at org.jpmml.sklearn.Main.main(Main.java:94)

-------------------------------------------------------------------------------------

The versions of the packages I'm using as reported by the debug=True are:

python: 3.7.3
sklearn: 0.20.3
sklearn.externals.joblib: 0.13.0
pandas: 0.24.2
sklearn_pandas: 1.8.0
sklearn2pmml: 0.44.0
java: 1.8.0_211

-------------------------------------------------

And finally the python code looks like this....

        eval_set = [(X_test, y_test_d[target_state])]
        pipe = PMMLPipeline([('xgb', xgb)])
        fit_params = dict(xgb__eval_set=eval_set,
                            xgb__early_stopping_rounds=30,
                            xgb__verbose=IS_VERBOSE)
        pipe.fit(X_train, y_train_d[target_state], **fit_params)

.... saving to PMML

sklearn2pmml(pipe, os.path.join(ALGO_DIR, state + '.pmml'), debug=True)

Thank you very much for your help.

**Comment by vruusmann** [comment]
> java.lang.IllegalArgumentException: |u1
> at numpy.core.NDArrayUtil$TypeDescriptor.read(NDArrayUtil.java:564)

It's the JPMML-SkLearn library, which complains about currently unsupported array descriptor - if I'm not mistaken, then "`|u1`" represents unsigned 1-byte integers.

Quick fix - change the data type from unsigned integers to signed integers.

**Comment by emilpop-au** [comment]
That worked perfectly.  I had to change to this

>>> pipe.fit(X_train, y_train_d[target_state].astype(int), **fit_params)

(y_train_d was created with pandas.get_dummies())

Thank you.

**Comment by vruusmann** [comment]
Reopening, because it would be a good idea to support more unsigned integer types (at least 1- and 2-byte ones).

And the exception message should be more informative ("Array data type <X> is not supported. Please change to types <Y> or <Z>", and "<A>-byte encoding of array data type <X> is not supported. Please change to <B> or <C>-byte encodings").

As shown here, this issue typically affects the label column.
